File: server/image_service.py
from PIL import Image

def adjust_image(source_path, target_path, output_path, source_img, subject_box, target_img, placement_point):
    # Calculate crop box
    crop_box = calculate_crop_box(source_img, subject_box, target_img)
    # Open images
    source_img_obj = Image.open(source_path)
    target_img_obj = Image.open(target_path)

    # Crop and paste
    cropped_subject = source_img_obj.crop((
        crop_box['x'],
        crop_box['y'],
        crop_box['x'] + crop_box['width'],
        crop_box['y'] + crop_box['height']
    ))
    # Adjust placement
    offset_x = subject_box['x'] - crop_box['x']
    offset_y = subject_box['y'] - crop_box['y']
    placement_box = {
        "x": placement_point['x'] - offset_x,
        "y": placement_point['y'] - offset_y,
        "width": crop_box['width'],
        "height": crop_box['height']
    }
    target_img_obj.paste(cropped_subject, (placement_box['x'], placement_box['y']))
    target_img_obj.save(output_path)

    return crop_box, placement_box

def calculate_crop_box(source_img, subject_box, target_img):
    subject_aspect = subject_box['width'] / subject_box['height']
    max_crop_width = min(target_img['width'], source_img['width'])
    max_crop_height = min(target_img['height'], source_img['height'])

    crop_width = max_crop_width
    crop_height = crop_width / subject_aspect
    if crop_height > max_crop_height:
        crop_height = max_crop_height
        crop_width = crop_height * subject_aspect

    subject_center_x = subject_box['x'] + subject_box['width'] / 2
    subject_center_y = subject_box['y'] + subject_box['height'] / 2
    crop_x = max(0, subject_center_x - crop_width / 2)
    crop_y = max(0, subject_center_y - crop_height / 2)

    # Stay within bounds
    if crop_x + crop_width > source_img['width']:
        crop_x = source_img['width'] - crop_width
    if crop_y + crop_height > source_img['height']:
        crop_y = source_img['height'] - crop_height

    return {
        "x": int(crop_x),
        "y": int(crop_y),
        "width": int(crop_width),
        "height": int(crop_height)
    }

File: src/ImageService.ts
import sharp from "sharp";

/**
 * Input parameters
 * - Paths to the source image, target image, and output image.
 * - Dimensions of the source image and target image.
 * - Subject box (area of interest in the source image) and placement point (where the subject will be placed on the target image).
 */
const sourceImagePath = "/Users/abhinavanand/Desktop/billi/source.png";
const targetImagePath = "/Users/abhinavanand/Desktop/billi/target.png";
const outputImagePath = "/Users/abhinavanand/Desktop/billi/output.png";

const sourceImage = { width: 796, height: 452 }; // Dimensions of the source image
const subjectBox = { x: 283, y: 124, width: 306, height: 320 }; // Area of interest in the source image
const targetImage = { width: 452, height: 796 }; // Dimensions of the target image
const placementPoint = { x: 72, y: 236 }; // Point on the target image where the subject will be placed

/**
 * Calculates the crop area from the source image and the placement area on the target image.
 * This ensures the subject is cropped and resized to fit the target image while maintaining its aspect ratio.
 */
function calculateCropAndPlacement() {
  const subjectAspectRatio = subjectBox.width / subjectBox.height; // Aspect ratio of the subject

  // Calculate the maximum possible crop dimensions that fit within both the source and target images
  const maxCropWidth = Math.min(targetImage.width, sourceImage.width);
  const maxCropHeight = Math.min(targetImage.height, sourceImage.height);

  let cropWidth = maxCropWidth;
  let cropHeight = cropWidth / subjectAspectRatio;

  // Adjust crop dimensions if the height exceeds the maximum allowed height
  if (cropHeight > maxCropHeight) {
    cropHeight = maxCropHeight;
    cropWidth = cropHeight * subjectAspectRatio;
  }

  // Center the crop area around the subject
  const subjectCenterX = subjectBox.x + subjectBox.width / 2;
  const subjectCenterY = subjectBox.y + subjectBox.height / 2;

  let cropX = subjectCenterX - cropWidth / 2;
  let cropY = subjectCenterY - cropHeight / 2;

  // Ensure the crop area stays within the bounds of the source image
  if (cropX < 0) cropX = 0;
  if (cropY < 0) cropY = 0;
  if (cropX + cropWidth > sourceImage.width)
    cropX = sourceImage.width - cropWidth;
  if (cropY + cropHeight > sourceImage.height)
    cropY = sourceImage.height - cropHeight;

  const cropBox = {
    x: Math.round(cropX),
    y: Math.round(cropY),
    width: Math.round(cropWidth),
    height: Math.round(cropHeight),
  };

  // Adjust the placement point on the target image based on the crop area
  const offsetX = subjectBox.x - cropBox.x;
  const offsetY = subjectBox.y - cropBox.y;

  const placementBox = {
    x: placementPoint.x - offsetX,
    y: placementPoint.y - offsetY,
    width: cropBox.width,
    height: cropBox.height,
  };

  console.log("Crop Box:", cropBox);
  console.log("Placement Box:", placementBox);

  return { cropBox, placementBox };
}

/**
 * Crops the subject from the source image based on the calculated crop area.
 * @param sourceImagePath Path to the source image
 * @param cropBox The area to crop from the source image
 * @returns A buffer containing the cropped image
 */
async function cropSubject(
  sourceImagePath: string,
  cropBox: { x: number; y: number; width: number; height: number }
): Promise<Buffer> {
  const meta = await sharp(sourceImagePath).metadata();

  // Adjust the crop box to ensure it stays within the bounds of the source image
  const adjustedCropBox = {
    left: Math.max(0, cropBox.x),
    top: Math.max(0, cropBox.y),
    width: Math.min(cropBox.width, (meta.width ?? 0) - cropBox.x),
    height: Math.min(cropBox.height, (meta.height ?? 0) - cropBox.y),
  };

  console.log("Adjusted Crop Box:", adjustedCropBox);

  // Extract the cropped area from the source image
  return sharp(sourceImagePath).extract(adjustedCropBox).toBuffer();
}

/**
 * Places the cropped subject onto the target image at the specified placement area.
 * @param sourceImagePath Path to the source image
 * @param targetImagePath Path to the target image
 * @param cropBox The area to crop from the source image
 * @param placementBox The area on the target image where the cropped subject will be placed
 * @param outputImagePath Path to save the final output image
 */
async function placeSubjectOnExistingTarget(
  sourceImagePath: string,
  targetImagePath: string,
  cropBox: { x: number; y: number; width: number; height: number },
  placementBox: { x: number; y: number; width: number; height: number },
  outputImagePath: string
) {
  // Crop the subject from the source image
  const subjectBuffer = await cropSubject(sourceImagePath, cropBox);

  // Resize the cropped subject to fit the placement area
  const resizedSubject = await sharp(subjectBuffer)
    .resize(placementBox.width, placementBox.height)
    .toBuffer();

  // Composite the resized subject onto the target image
  await sharp(targetImagePath)
    .composite([
      {
        input: resizedSubject,
        top: placementBox.y,
        left: placementBox.x,
      },
    ])
    .toFile(outputImagePath);

  console.log(`Output image saved at: ${outputImagePath}`);
}

// Run the process
const { cropBox, placementBox } = calculateCropAndPlacement();
placeSubjectOnExistingTarget(
  sourceImagePath,
  targetImagePath,
  cropBox,
  placementBox,
  outputImagePath
);
